File: bindings/perl/Champlain/examples/capitals.pl
#!/usr/bin/perl

=head1 NAME

capitals.pl - Show the capitals

=head1 DESCRIPTION

This program takes you into a magical trip and displays the capital cities one
by one.

=cut

use strict;
use warnings;
use open ':std', ':utf8';

use Glib qw(TRUE FALSE);
use Clutter qw(-gtk-init);
use Gtk2 qw(-init);
use Champlain;

use XML::LibXML;


exit main();


sub main {

	my $window = Gtk2::Window->new();
	my $vbox = Gtk2::VBox->new(FALSE, 0);
	
	
	# Create the map stuff
	my $map = Champlain::View->new();
	my $gtk2_map = Gtk2::Champlain::ViewEmbed->new($map);
	$gtk2_map->set_size_request(640, 480);
	$map->center_on(0, 0);
	$map->set('scroll-mode', 'kinetic', 'zoom-level', 3);
	
	my $layer = Champlain::Layer->new();
	$map->add_layer($layer);
	
	
	my $viewport = Gtk2::Viewport->new();
	$viewport->set_shadow_type('etched-in');
	$viewport->add($gtk2_map);
	$vbox->pack_start($viewport, TRUE, TRUE, 0);
	
	$window->add($vbox);
	$window->set_size_request($gtk2_map->get_size_request);
	$window->signal_connect(destroy => sub {
		Gtk2->main_quit();
	});
	$window->show_all();

	
	my $capitals_url = "http://en.wikipedia.org/wiki/List_of_national_capitals";
	my $soup = My::Soup->new($capitals_url);
	
	$soup->do_get(
		$capitals_url,
		\&capitals_main_callback,
		{
			map   => $map,
			layer => $layer,
			markers => [],
		}
	);
	
	
	Gtk2->main();
	
	
	return 0;
}


#
# Called when the main page with all the capitals is downloaded.
#
sub capitals_main_callback {
	my ($soup, $uri, $response, $data) = @_;
	
	my $parser = XML::LibXML->new();
	$parser->recover_silently(1);
	$data->{parser} = $parser;

	# Find the table with the capitals
	my $document = $parser->parse_html_string($response->content);
	my @nodes = $document->findnodes('//table[@class="wikitable sortable"]/tr/td[1]/a');
	
	# Get the capitals
	my @capitals = ();
	foreach my $node (@nodes) {
		my $uri = $node->getAttribute('href') or next;
		my $name = $node->getAttribute('title') or next;
		my $capitals = {
			uri => $uri,
			name => $name,
		};
		push @capitals, $capitals;
	}
	
	# Download the capitals (the download is node one capital at a time)
	$data->{capitals} = \@capitals;
	download_capital($soup, $data);
}


#
# Called when the page of a capital is downloaded. The page is expected to have
# the coordinates of the capital.
#
sub capital_callback {
	my ($soup, $uri, $response, $data) = @_;
	
	my $document = $data->{parser}->parse_html_string($response->content);
	my $heading = $document->getElementById('firstHeading');
	if ($heading) {
		my $name = $document->getElementById('firstHeading')->textContent;
				
		my ($geo) = $document->findnodes('id("coordinates")//span[@class="geo"]');
		if ($geo) {
			my ($latitude, $longitude) = split /\s*;\s*/, $geo->textContent;
			$data->{current}{latitude} = $latitude;
			$data->{current}{longitude} = $longitude;
			printf "$name %.4f, %.4f\n", $latitude, $longitude;
	
			my $marker = Champlain::Marker->new_with_label($name, "Sans 15", undef, undef);
			$marker->set_position($latitude, $longitude);
			if (@{ $data->{markers} } == 5) {
				my $old = shift @{ $data->{markers} };
				$data->{layer}->remove($old);
			}
			push @{ $data->{markers} }, $marker;
			$data->{layer}->add($marker);
			$data->{map}->go_to($latitude, $longitude);
		}
	}
	
	Glib::Timeout->add (1_000, sub {
		download_capital($soup, $data);
		return FALSE;
	});
}


#
# This function downloads the page of a capital and then call it self again with
# the next capital to download. This process is repeated until there are no more
# capitals to download.
#
# The capitals to download are taken from $data->{capitals}.
#
sub download_capital {
	my ($soup, $data) = @_;

	my $capital = shift @{ $data->{capitals} };
	if (! defined $capital) {
		print "No more capitals to download\n";
		return;
	}
	
	my $uri = $capital->{uri};
	my $name = $capital->{name};
	$data->{current} = $capital;
	$soup->do_get($uri, \&capital_callback, $data);
}



#
# A very cheap implementation of an asynchornous HTTP client that integrates
# with Glib's main loop. This client implements a rudimentary version of
# 'Keep-Alive'.
#
# Each instance of this class can only make HTTP GET requests and only to a
# single HTTP server.
#
#
# Usage:
#
#   my $soup = My::Soup->new('http://en.wikipedia.com/');
#   $soup->do_get('http://en.wikipedia.com/Bratislava', sub {
#     my ($soup, $uri, $response, $data) = @_;
#     print $response->content;
#   });
#
package My::Soup;

use Glib qw(TRUE FALSE);
use Net::HTTP::NB;
use HTTP::Response;
use URI;


sub new {
	my $class = shift;
	my ($uri) = @_;
	
	my $self = bless {}, ref $class || $class;

	$uri = to_uri($uri);
	$self->{port} = $uri->port;
	$self->{host} = $uri->host;
	
	$self->connect();
	
	return $self;
}


#
# Connects to the remote HTTP server.
#
sub connect {
	my $self = shift;
	my $http = Net::HTTP::NB->new(
		Host      => $self->{host},
		PeerPort  => $self->{port},
		KeepAlive => 1,
	);
	$self->http($http);
}


sub http {
	my $self = shift;
	if (@_) {
		$self->{http} = $_[0];
	}
	return $self->{http};
}


sub to_uri {
	my ($uri) = @_;
	return $uri if ref($uri) && $uri->isa('URI');
	return URI->new($uri);
}


#
# Performs an HTTP GET request asynchronously.
#
sub do_get {
	my $self = shift;
	my ($uri, $callback, $data) = @_;
	$uri = to_uri($uri);
	
	# Note that this is not asynchronous!
	$self->http->write_request(GET => $uri->path_query);
	
	
	my ($code, $message, %headers);
	my $content = "";
	Glib::IO->add_watch($self->http->fileno, ['in'], sub {
		my (undef, $condition) = @_;
		
		# Read the headers
		if (!$code) {
			eval {
				($code, $message, %headers) = $self->http->read_response_headers();
			};
			if (my $error = $@) {
				# The server closed the socket reconnect and resume the HTTP GET
				$self->connect();
				$self->do_get($uri, $callback, $data);
				
				# We abort this I/O watch since another download will be started
				return FALSE;
			}
			
			# We return and continue when the server will have more data
			return TRUE;
		}
		
		
		# Read the content		
		my $line;
		my $n = $self->http->read_entity_body($line, 1024);
		$content .= $line;
		
		if ($self->http->keep_alive) {
			# In the case where the HTTP request has keep-alive we need to see if the
			# content has all arrived as read_entity_body() will not tell when the end
			# of the content has been reached.
			return TRUE unless length($content) == $headers{'Content-Length'};
		}
		elsif ($n) {
			# There's still data to read
			return TRUE;
		}
		
		# End of the document
		my $response = HTTP::Response->new($code, $message, [%headers], $content);
		$callback->($self, $uri, $response, $data);
		return FALSE;
	});
}

# A true value
1;

File: bindings/python/demos/capitals.py
#!/usr/bin/env python

import clutter, cluttergtk
import champlain
import gtk, gobject
from twisted.internet import gtk2reactor
gtk2reactor.install()
from twisted.internet import reactor
import twisted.web.client as httpclient
from BeautifulSoup import BeautifulSoup


BASE_URL = "http://en.wikipedia.org"
CAPITALS_URL = "%s/wiki/List_of_national_capitals" % BASE_URL

class Capitals:


	def __init__(self):
		self.win = gtk.Window()
		self.win.set_title("A Capital Idea")

		self.capital_uris = []
		self.markers = []
	
		# Create the map stuff
		self.map = champlain.View()
		self.embed = cluttergtk.Embed()
		self.embed.set_size_request(640, 480)

		self.map.set_property("scroll-mode", champlain.SCROLL_MODE_KINETIC)
		self.map.set_property("zoom-level", 3)

		self.layer = champlain.Layer()
		self.map.add_layer(self.layer)

		self.win.add(self.embed)
		self.win.connect("destroy", lambda w: reactor.stop())

		self.embed.realize()
		stage = self.embed.get_stage()
		self.map.set_size(640, 480)
		stage.add(self.map)

		self.win.show_all()
		self.map.center_on(0, 0)

		# Download the next map after the go-to animation has been completed
		self.map.connect('animation-completed::go-to', lambda w: gobject.timeout_add_seconds(1, self.download_capital))
	
		d = httpclient.getPage(CAPITALS_URL)
		d.addCallback(self.capitals_main_cb)


	def capitals_main_cb(self, data):
		"""
		Called when the main page with all the capitals is downloaded.
		"""

		soup = BeautifulSoup(data)
		table = soup.find("table", { "class" : "wikitable sortable" })
		for row in table.findAll("tr"):
			cell = row.find("td")
			if cell:
				link = cell.find("a")
				uri = str(link['href'])
				self.capital_uris.append(uri)

		self.download_capital()


	def capital_cb(self, data):
		"""
		Called when the page of a capital is downloaded. The page is expected to have
		the coordinates of the capital.
		"""

		soup = BeautifulSoup(data)
		heading = soup.find("h1", { "id" : "firstHeading" })
		if not heading:
			return
		name = heading.contents[0]

		geo = soup.find("span", { "class" : "geo" })
		if not geo:
			return
		latitude, longitude = geo.contents[0].split(";")
		latitude = float(latitude)
		longitude = float(longitude)

		# Keep only a few capitals at each iteration we remove a capital
		if len(self.markers) == 5:
			marker = self.markers[0]
			self.layer.remove(marker)
			self.markers.remove(marker)

		font = "Sans 15"
		white = clutter.Color(0xff, 0xff, 0xff, 0xff)
		orange = clutter.Color(0xf3, 0x94, 0x07, 0xbb)
		black = clutter.Color(0x00, 0x00, 0x00, 0xff)

		if self.markers:
			#Change the colour of the last marker's text
			last = self.markers.pop()
			self.layer.remove(last)
			marker = champlain.marker_new_with_text(last.name, font, black, orange)
			marker.set_position(last.get_property("latitude"), last.get_property("longitude"))
			self.markers.append(marker)
			self.layer.add(marker)
			marker.raise_top()

		marker = champlain.marker_new_with_text(name, font, white, orange)
		marker.set_position(latitude, longitude)
		marker.name = name
		self.markers.append(marker)
		self.layer.add(marker)
		marker.raise_top()
		self.map.go_to(latitude, longitude)


	def download_capital(self):
		if not self.capital_uris:
			# No more capitals to display
			return
		uri = self.capital_uris[0]
		self.capital_uris.remove(uri)
		d = httpclient.getPage(BASE_URL + uri)
		d.addCallback(self.capital_cb)


if __name__ == "__main__":
	gobject.threads_init()
	clutter.init()
	Capitals()
	reactor.run()
